wvd-templates/DSC/Functions.ps1

```diff
@@ -327,4 +330,163 @@ function IsRDAgentRegistryValidForRegistration {
+    
+        [Parameter(Mandatory = $false)]
+        [ValidateNotNullOrEmpty()]
+        [bool]$StartAgent,
+    
+        [Parameter(mandatory = $false)]
+        [switch]$EnableVerboseMsiLogging
+    )
+
+    # Convert relative paths to absolute paths if needed
+    Write-Log -Message "Boot loader folder is $AgentBootServiceInstallerFolder"
+    $AgentBootServiceInstaller = (Get-ChildItem $AgentBootServiceInstallerFolder\ -Filter *.msi | Select-Object).FullName
+    if ((-not $AgentBootServiceInstaller) -or (-not (Test-Path $AgentBootServiceInstaller))) {
+        throw "RD Infra Agent Installer package is not found '$AgentBootServiceInstaller'"
+    }
+
+    # Convert relative paths to absolute paths if needed
+    Write-Log -Message "Agent folder is $AgentInstallerFolder"
+    $AgentInstaller = (Get-ChildItem $AgentInstallerFolder\ -Filter *.msi | Select-Object).FullName
+    if ((-not $AgentInstaller) -or (-not (Test-Path $AgentInstaller))) {
+        throw "RD Infra Agent Installer package is not found '$AgentInstaller'"
+    }
+
+    if (!$RegistrationToken) {
+        throw "No registration token specified"
+    }
+
+    RunMsiWithRetry -programDisplayName "RDAgentBootLoader" -isUninstall -argumentList @("/x {A38EE409-424D-4A0D-B5B6-5D66F20F62A5}", "/quiet", "/qn", "/norestart", "/passive") -msiOutputLogPath "C:\Users\AgentBootLoaderUnInstall.txt" -msiLogVerboseOutput:$EnableVerboseMsiLogging
+
+    while ($true) {
+        try {
+            $oldAgent = Get-Package -ProviderName msi -Name "Remote Desktop Services Infrastructure Agent" -ErrorAction Stop
+        }
+        catch {
+            #Ignore the error if it was due to no packages being found.
+            if ($_.FullyQualifiedErrorId -eq "NoMatchFound,Microsoft.PowerShell.PackageManagement.Cmdlets.GetPackage") {
+                break
+            }
+
+            throw;
+        }
+
+        $oldVersion = $oldAgent.Version
+        $productCodeParameter = $oldAgent.FastPackageReference
+
+        RunMsiWithRetry -programDisplayName "RD Infra Agent $oldVersion" -isUninstall -argumentList @("/x $productCodeParameter", "/quiet", "/qn", "/norestart", "/passive") -msiOutputLogPath "C:\Users\AgentUninstall.txt" -msiLogVerboseOutput:$EnableVerboseMsiLogging
+    }
+
+
+    Write-Log -Message "Installing RD Infra Agent on VM $AgentInstaller"
+    RunMsiWithRetry -programDisplayName "RD Infra Agent" -argumentList @("/i $AgentInstaller", "/quiet", "/qn", "/norestart", "/passive", "REGISTRATIONTOKEN=$RegistrationToken") -msiOutputLogPath "C:\Users\AgentInstall.txt" -msiLogVerboseOutput:$EnableVerboseMsiLogging
+
+    Write-Log -Message "Installing RDAgent BootLoader on VM $AgentBootServiceInstaller"
+    RunMsiWithRetry -programDisplayName "RDAgent BootLoader" -argumentList @("/i $AgentBootServiceInstaller", "/quiet", "/qn", "/norestart", "/passive") -msiOutputLogPath "C:\Users\AgentBootLoaderInstall.txt" -msiLogVerboseOutput:$EnableVerboseMsiLogging
+
+    if ($StartAgent) {

```

Comment:
If installation failed then no point in attempting to start agent, might want to throw an error to fail fast instead of DSC to timeout eventually, because there will be a lot of polling after this.

Reply:
Good point, I will throw an exception if the RDAgentBootLoader retry limit is exceeded

Reply:
Regarding checking if the installation failed the code this is replacing didn't do that, so this has the same behavior except that it does throw an exception if it fails with exit code 1618 after a certain number of retries. I don't think we should try to find all of the error codes considered successful or not as part of this change since per https://docs.microsoft.com/en-us/windows/win32/msi/error-codes since there are multiple error codes that appear to means successful or not applicable. If we want to make that change we should do it in a separate internal task.